File: java/src/main/java/com/fleetmgr/sdk/client/state/pilot/connected/ValidatingChannels.java
package com.fleetmgr.sdk.client.state.pilot.connected;

import com.fleetmgr.interfaces.ChannelIndicationList;
import com.fleetmgr.interfaces.ChannelResponse;
import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.connection.Received;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.traffic.Channel;
import com.fleetmgr.sdk.client.traffic.ChannelImpl;
import com.fleetmgr.sdk.client.event.output.facade.ChannelsOpened;
import com.fleetmgr.sdk.client.state.State;
import com.fleetmgr.interfaces.Role;
import com.fleetmgr.interfaces.facade.control.ClientMessage;
import com.fleetmgr.interfaces.facade.control.Command;
import com.fleetmgr.interfaces.facade.control.ControlMessage;
import com.fleetmgr.interfaces.facade.control.Response;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Created by: Bartosz Nawrot
 * Date: 24.09.2018
 * Description:
 */
public class ValidatingChannels extends State {

    private Role role;
    private List<ChannelResponse> channels;

    private Map<Long, Channel> validated;

    public ValidatingChannels(State state, Role role, List<ChannelResponse> channels) {
        super(state);
        this.role = role;
        this.channels = channels;
    }

    @Override
    public State start() {
        validated = backend.getChannelsHandler().validateChannels(channels);
        send(ClientMessage.newBuilder()
                .setCommand(Command.CHANNELS_READY)
                .setChannelsIndication(ChannelIndicationList.newBuilder()
                        .addAllIds(validated.keySet())
                        .build())
                .build());
        return null;
    }

    @Override
    public State notifyEvent(UserEvent event) {
        switch (event.getType()) {
            case RELEASE:
                return new Releasing(this);

            default:
                return defaultEventHandle(event.toString());
        }
    }

    @Override
    public State notifyConnection(ConnectionEvent event) {
        switch (event.getType()) {
            case RECEIVED:
                return handleMessage(((Received) event).getMessage());

            default:
                return defaultEventHandle(event.toString());
        }
    }

    private State handleMessage(ControlMessage message) {
        switch (message.getCommand()) {
            case CHANNELS_READY:
                if (message.getResponse() == Response.ACCEPTED) {
                    listener.onEvent(new ChannelsOpened(new LinkedList<>(validated.values())));
                    switch (role) {
                        case LEADER:
                            return new Controlling(this);
                        case PILOT:
                            return new Spectating(this);
                    }

                } else {
                    return defaultMessageHandle(message);
                }

            default:
                return defaultMessageHandle(message);
        }
    }

    @Override
    public String toString() {
        return "ValidatingChannels";
    }
}

File: cpp/source/state/pilot/connected/ValidatingChannels.cpp
#include "state/pilot/connected/ValidatingChannels.hpp"

#include "state/pilot/connected/Controlling.hpp"
#include "state/pilot/connected/Spectating.hpp"

#include "event/input/connection/Received.hpp"

#include "event/output/ChannelsOpened.hpp"
#include "event/output/Error.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::pilot;
using namespace fm::state::pilot::connected;

using namespace com::fleetmgr::interfaces::facade::control;

using event::input::user::UserEvent;
using event::input::connection::ConnectionEvent;
using event::input::connection::Received;

using event::output::FacadeEvent;
using event::output::ChannelsOpened;
using event::output::Error;

using com::fleetmgr::interfaces::Role;
using com::fleetmgr::interfaces::Channel;

ValidatingChannels::ValidatingChannels(IState& state, Role _role, const std::vector<Channel>& openedChannels) :
    IState(state),
    role(_role),
    toValidate(openedChannels)
{
}

std::unique_ptr<IState> ValidatingChannels::start()
{
    auto result = backend.validateChannels(toValidate);
    ClientMessage response;
    response.set_command(Command::CHANNELS_READY);
    for (auto& pair : result)
    {
        response.mutable_attachchannels()->add_attachedchannels(pair.first);
    }
    send(response);
    return nullptr;
}

std::string ValidatingChannels::toString() const
{
    return "ValidatingChannels";
}

std::unique_ptr<IState> ValidatingChannels::handleUserEvent(const UserEvent& event)
{
    switch (event.getType())
    {
    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> ValidatingChannels::handleConnectionEvent(const ConnectionEvent& event)
{
    switch (event.getType())
    {
    case ConnectionEvent::RECEIVED:
        return handleMessage(reinterpret_cast<const Received&>(event).getMessage());

    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> ValidatingChannels::handleMessage(const ControlMessage& message)
{
    switch (message.command())
    {
    case Command::CHANNELS_READY:
        if (message.response() == Response::ACCEPTED)
        {
            listener.onEvent(std::make_shared<ChannelsOpened>());
            switch (role)
            {
            case Role::LEADER:
                return std::make_unique<Controlling>(*this);

            case Role::PILOT:
                return std::make_unique<Spectating>(*this);
            }
        }
        else
        {
            listener.onEvent(std::make_shared<Error>());
            return defaultMessageHandle(message);
        }

    default:
        return defaultMessageHandle(message);
    }
}
